drivers/gpio/gpio_rpi_pico.c

```diff
@@ -69,28 +89,62 @@ static int gpio_rpi_port_get_raw(const struct device *dev, uint32_t *value)
 static int gpio_rpi_port_set_masked_raw(const struct device *port,
 					uint32_t mask, uint32_t value)
 {
-	gpio_put_masked(mask, value);
+	struct gpio_rpi_data *data = port->data;
+	/* First handle push-pull pins: */
+	gpio_put_masked(mask & ~data->single_ended_mask, value);
+	/* Then handle open-drain pins: */
+	gpio_set_dir_masked(mask & data->single_ended_mask & data->open_drain_mask, ~value);
+	/* Then handle open-source pins: */
+	gpio_set_dir_masked(mask & data->single_ended_mask & ~data->open_drain_mask, value);
 	return 0;
 }
 
 static int gpio_rpi_port_set_bits_raw(const struct device *port,
 					uint32_t pins)
 {
-	gpio_set_mask(pins);
+	struct gpio_rpi_data *data = port->data;
+	/* First handle push-pull pins: */
+	gpio_set_mask(pins & ~data->single_ended_mask);
+	/* Then handle open-drain pins: */
+	gpio_set_dir_in_masked(pins & data->single_ended_mask & data->open_drain_mask);
+	/* Then handle open-source pins: */
+	gpio_set_dir_out_masked(pins & data->single_ended_mask & ~data->open_drain_mask);
 	return 0;
 }
 
 static int gpio_rpi_port_clear_bits_raw(const struct device *port,
 					uint32_t pins)
 {
-	gpio_clr_mask(pins);
+	struct gpio_rpi_data *data = port->data;
+	/* First handle push-pull pins: */
+	gpio_clr_mask(pins & ~data->single_ended_mask);
+	/* Then handle open-drain pins: */
+	gpio_set_dir_out_masked(pins & data->single_ended_mask & data->open_drain_mask);
+	/* Then handle open-source pins: */
+	gpio_set_dir_in_masked(pins & data->single_ended_mask & ~data->open_drain_mask);
 	return 0;
 }
 
 static int gpio_rpi_port_toggle_bits(const struct device *port,
 					uint32_t pins)
 {
-	gpio_xor_mask(pins);
+	struct gpio_rpi_data *data = port->data;
+	/* First handle push-pull pins: */
+	gpio_xor_mask(pins & ~data->single_ended_mask);
+	/* Then handle single-ended pins: */
+#ifdef sio_hw
+	/* (unfortunately there's no pico-sdk api call that can be used for this,
+	 * but it's possible by accessing the registers directly)
+	 */
+	sio_hw->gpio_oe_togl = (pins & data->single_ended_mask);
```

Comment:
You can keep this line and remove the ifdef. It's unlikely that we will lose access to the registers. If it will ever happen, we will change this accordingly

Reply:
DONE